### PR TITLE
Manage thrift thread-pool and interrupt it during shutdown

### PR DESCRIPTION
**Pull request by ggezer**
Thrift server doesn't interrupt its `ExecutorService` during shutdown. This PR makes it use an external `ExecutorService` that we own, and interrupt it after shutting down Thrift.

### CODE REVIEW

**Comment by gpang** [comment]
@ggezer  this should target master, then cherry-picked to older branches.

Also, could you add a description?

**Comment by ggezer** [comment]
@gpang This fix is not applicable to master.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 1]
```diff
@@ -382,16 +388,22 @@ protected void startServingRPCServer() {
     } catch (TTransportException e) {
       throw new RuntimeException(e);
     }
+    // create executor service for thrift.
+    // use shutdown timeout as keep-alive.
+    int stopTimeoutMs = (int) Configuration.getMs(PropertyKey.MASTER_THRIFT_SHUTDOWN_TIMEOUT);
+    mExecutorService = new ThreadPoolExecutor(mMinWorkerThreads,
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: is this similar to how thrift would have created its own executor service internally?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 2]
```diff
@@ -382,16 +388,22 @@ protected void startServingRPCServer() {
     } catch (TTransportException e) {
       throw new RuntimeException(e);
     }
+    // create executor service for thrift.
+    // use shutdown timeout as keep-alive.
+    int stopTimeoutMs = (int) Configuration.getMs(PropertyKey.MASTER_THRIFT_SHUTDOWN_TIMEOUT);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: NIT: rename to `shutdownTimeoutMs`

Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: I think. stop would make more sense as it's what thrift calls it. Changed the doc.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 1 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: It's functionally equivalent to how thrift creates internally.

**Comment by ggezer** [comment]
Jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 3]
```diff
@@ -406,7 +418,11 @@ protected void startServingRPCServer() {
   protected void stopServing() throws Exception {
     if (mThriftServer != null) {
       mThriftServer.stop();
+      // Thrift does not interrupt outstanding tasks.
+      // Call shutdownNow() to interrupt them.
+      mExecutorService.shutdownNow();
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: @ggezer  Is there any way to detect if all the threads are "done" or not?

Comment (calvinjia) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Can we wait for a period of time to ensure all threads are stopped and the threadpool is no longer accepting tasks? If there are lingering threads still I think it is safe to crash.

Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Thrift already calls shutdown and waits for a configured timeout before returning. Problem is that it doesn't to anything for lingering tasks. So this call will just make sure to terminate terminate them after grace period.

Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: If there is a way to detect if threads are still alive, then we should "exit", right? That would ensure that we start the server without any lingering threads.

Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: @gpang Are you suggesting a second level of timeout? At the point we called `mExecutor.shutdownNow()` we already waited upto `stopTimeout` for outstanding tasks to finish. `shutdownNow` just interrupts if there any existing threads. After calling `shutdownNow` there is no way to block on completion. We could only see if there are any before invoking it.

Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: @ggezer what about `awaitTermination` ? https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination(long,%20java.util.concurrent.TimeUnit)

would we be able to detect with that?

**Comment by ggezer** [comment]
When I say thrift awaits after calling `shutdown`, it's by calling `awaitTermination`. After `shutdownNow` is called, `awaitTermination` returns immediately.

**Comment by gpang** [comment]
@ggezer The `ThreadPoolExecutor#shutdownNow()` javadoc says: `This method does not wait for actively executing tasks to terminate.  Use awaitTermination to do that.` Doesn't that mean we would still have to call `awaitTermination()` after `shutdownNow()` to ensure running threads are completed?

**Comment by gpang** [comment]
alluxio-bot, merge this please

**Pull request merged**